### PR TITLE
[WIP] - Cleanup

### PR DESCRIPTION
**Pull request by jcrist**
Trying to get an organized and maintainable structure. This is mostly just merging things in from other working branches in a combined and sensible way.

### CODE REVIEW

File: dklearn/wrappers.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+
+class DelayedEstimator(Base):
+    """A delayed estimator"""
+    __slots__ = ('key', 'dask', '_base')
+    _default_get = staticmethod(threaded.get)
+    _optimize = staticmethod(lambda dsk, keys, **kwargs: dsk)
+
+    @property
+    def _finalize(self):
+        return partial(_finalize_estimator, base=self._base)
```
Comment (mrocklin) [inline dklearn/wrappers.py]: Is it possible to make finalize a staticmethod or classmethod?

File: dklearn/wrappers.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+        return self._to_delayed(res)
+
+    @if_delegate_has_method('estimator')
+    def predict(self, X, compute=True):
+        func = partial(predict_chunk, self.estimator)
+        if isinstance(X, da.Array):
+            assert X.ndim == 2
+            if len(X.chunks[1]) != 1:
+                X = X.rechunk((X.chunks[0], X.shape[1]))
+            res = X.map_blocks(func, chunks=(X.chunks[0],), drop_axis=1)
```
Comment (mrocklin) [inline dklearn/wrappers.py]: map_blocks may now take kwargs, allowing you to drop the partial

File: dklearn/wrappers.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+        elif isinstance(X, db.Bag):
+            res = X.map_partitions(func)
+        elif is_list_of(X, Delayed):
+            func = delayed(func, pure=True)
+            res = [func(i) for i in X]
+            if compute:
+                return list(dask.compute(*res))
+            return res
+        else:
+            return func(X) if compute else delayed(func, pure=True)(X)
```
Comment (mrocklin) [inline dklearn/wrappers.py]: Thoughts on always normalizing to lists of delayed?  I guess this is bad if you want to keep the input type.

File: dklearn/wrappers.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+        return res.compute() if compute else res
+
+    @if_delegate_has_method('estimator')
+    def score(self, X, y, sample_weight=None, compute=True):
+        if compute:
+            X, y, sample_weight = dask.compute(X, y, sample_weight)
+            return self.estimator.score(X, y, sample_weight=sample_weight)
```
Comment (mrocklin) [inline dklearn/wrappers.py]: I would have expected to see this task also injected into the graph

File: dklearn/wrappers.py [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+        x_parts, y_parts = as_lists_of_delayed(X, y)
+        est = self.estimator
+        chunks = [fit_chunk(est, xp, yp, **kwargs) for (xp, yp) in
+                  zip(x_parts, y_parts)]
+        return average_coefs(chunks)
+
+
+@delayed(pure=True)
+def _unique_chunk(x):
+    return np.unique(x)
```
Comment (mrocklin) [inline dklearn/wrappers.py]: Why the alias?  Does `np.unique` not serialize well?

File: dklearn/wrappers.py [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+
+
+@delayed(pure=True)
+def _unique_merge(x):
+    return np.unique(np.concatenate(x))
+
+
+@delayed(pure=True)
+def partial_fit_chunk(est, x, y, **kwargs):
+    return est.partial_fit(x, y, **kwargs)
```
Comment (mrocklin) [inline dklearn/wrappers.py]: This may not be pure

File: dklearn/wrappers.py [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+        n_features = estimators[0].coef_.shape[1]
+        coef = np.zeros((n_classes, n_features), dtype='f8')
+        intercept = np.zeros(n_classes, dtype='f8')
+        for m in estimators:
+            ind = np.in1d(classes, m.classes_)
+            coef[ind] += m.coef_
+            intercept[ind] += m.intercept_
+        o.coef_ = coef / len(estimators)
+        o.intercept_ = intercept / len(estimators)
+    return o
```
Comment (mrocklin) [inline dklearn/wrappers.py]: Are there different ways to average results?  Does sklearn have things we should steal here?

File: dklearn/wrappers.py (hunk 1 above)
Comment (jcrist) [inline dklearn/wrappers.py]: It's a property that returns a partial containing a reference to the base class on which it needs to be assigned. There is no reference to the parent object here, it's effectively the same as a staticmethod.

File: dklearn/wrappers.py (hunk 3 above)
Comment (jcrist) [inline dklearn/wrappers.py]: My thought is wherever possible type in => type out. Lists of delayed and bags lose metadata, but are more general - arrays are best.

File: dklearn/wrappers.py (hunk 5 above)
Comment (jcrist) [inline dklearn/wrappers.py]: No real reason, could do `_unique_chunk = delayed(np.unique, pure=True)` just as well.

File: dklearn/wrappers.py (hunk 7 above)
Comment (jcrist) [inline dklearn/wrappers.py]: Afaik scikit-learn doesn't implement this. I'm actually unsure if the missing classes case is correct (I think it is?).

File: dklearn/wrappers.py (hunk 4 above)
Comment (jcrist) [inline dklearn/wrappers.py]: Since it's the final step in a reduction, this is effectively the same thing. I suppose this may matter if it's distributed and the worker->worker bandwidth/memory is higher than worker->client.

File: dklearn/wrappers.py (hunk 6 above)
Comment (jcrist) [inline dklearn/wrappers.py]: Good point.

File: dklearn/wrappers.py [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+
+
+def _finalize_estimator(res, base):
+    base = clone(base)
+    if hasattr(base, '_update_estimator'):
+        return base._update_estimator(res[0])
+    return copy_to(res[0], base)
+
+
+class DelayedEstimator(Base):
```
Comment (mrocklin) [inline dklearn/wrappers.py]: I'm still somewhat in favor of removing this class if we can do so cleanly, instead going with an xarray style of holding on to lazy results.  This preference isn't based on particularly deep experience though.  Thoughts?

File: dklearn/wrappers.py (hunk 3 above)
Comment (mrocklin) [inline dklearn/wrappers.py]: Is this pattern common enough to pull out to a separate map_blocks function that works on any of the input types?  I'm trying to reduce cognitive load for sklearn devs as they review this.

File: dklearn/wrappers.py (hunk 4 above)
Comment (mrocklin) [inline dklearn/wrappers.py]: Or if we care about profiling.  It also seems cleaner to me to say something like the following:

```python
result = score_chunk(self.estimator, X, y, sample_weight)
if compute:
    result = result.compute()
return result
```

There are fewer places where things might go wrong.

File: dklearn/wrappers.py (hunk 7 above)
Comment (mrocklin) [inline dklearn/wrappers.py]: From where did this approach originate?  Do we have a reference on why/when it works well?

File: dklearn/wrappers.py (hunk 5 above)
Comment (mrocklin) [inline dklearn/wrappers.py]: This would come through prettier in diagnostics like the progress.  

There is also some cleanliness value, I think, in just using numpy/pandas functions when they serialize well.

**Comment by mrocklin** [comment]
I suspect that combining these meta-estimators with pipeline/gridsearch will help to teach us about how to effectively nest parallelism
